### PR TITLE
[MPS] Migrate softmax to metal kernels

### PR DESCRIPTION
**Pull request by Isalia20**
Migrate softmax to metal kernels reusing log_softmax kernels. Perf (fp32 mainly in benchmarks because Transformers standard sampling path casts logits to fp32 before softmax):

| Shape | Dtype | Dim | Layout | MPSGraph (us) | Metal (us) | Speedup |
|---|---|---:|---|---:|---:|---:|
| `1x128256` | fp32 | -1 | contiguous | 57.37 | 7.14 | 8.035x |
| `1x129280` | fp32 | -1 | contiguous | 50.05 | 7.09 | 7.060x |
| `1x151936` | fp32 | -1 | contiguous | 56.22 | 7.58 | 7.420x |
| `1x154880` | fp32 | -1 | contiguous | 60.75 | 8.02 | 7.576x |
| `1x201088` | fp32 | -1 | contiguous | 74.11 | 9.68 | 7.656x |
| `4x128256` | fp32 | -1 | contiguous | 54.84 | 16.02 | 3.423x |
| `4x129280` | fp32 | -1 | contiguous | 54.42 | 16.30 | 3.339x |
| `4x151936` | fp32 | -1 | contiguous | 61.56 | 19.25 | 3.197x |
| `4x154880` | fp32 | -1 | contiguous | 67.01 | 19.32 | 3.468x |
| `4x201088` | fp32 | -1 | contiguous | 84.67 | 26.62 | 3.181x |
| `8x128256` | fp32 | -1 | contiguous | 63.15 | 33.93 | 1.862x |
| `8x129280` | fp32 | -1 | contiguous | 57.45 | 34.94 | 1.644x |
| `8x151936` | fp32 | -1 | contiguous | 65.42 | 39.73 | 1.646x |
| `8x154880` | fp32 | -1 | contiguous | 67.61 | 41.19 | 1.641x |
| `8x201088` | fp32 | -1 | contiguous | 80.48 | 50.81 | 1.584x |
| `5x51866` | fp32 | -1 | contiguous | 29.44 | 10.54 | 2.793x |
| `1x151936` | fp32 | -1 | offset view | 56.06 | 7.64 | 7.334x |
| `4x499x46` | fp32 | -1 | contiguous | 18.95 | 4.48 | 4.228x |
| `64x1000` | fp32 | -1 | contiguous | 20.37 | 8.40 | 2.424x |
| `4096x30522` | fp32 | -1 | contiguous | 5783.57 | 5515.95 | 1.049x |
| `2x150x512x512` | fp32 | 1 | contiguous | 5223.75 | 2989.85 | 1.747x |
| `4x499x2341` | fp32 | -1 | contiguous | 251.68 | 165.52 | 1.520x |
| `1024x50257` | fp32 | -1 | contiguous | 2602.69 | 3090.39 | 0.842x |
| `2048x128256` | fp32 | -1 | contiguous | 12540.65 | 15957.20 | 0.786x |
| `2048x129280` | fp32 | -1 | contiguous | 12311.65 | 16219.90 | 0.759x |
| `2048x151936` | fp32 | -1 | contiguous | 15902.57 | 18924.41 | 0.840x |
| `8192x151936` | fp32 | -1 | contiguous | 66602.24 | 78375.70 | 0.850x |
| `2048x154880` | fp32 | -1 | contiguous | 14568.02 | 20323.50 | 0.717x |
| `1024x201088` | fp32 | -1 | contiguous | 9858.95 | 12798.42 | 0.770x |
| `4x1023x151936` | bf16 | -1 | contiguous | 18029.44 | 20589.93 | 0.876x |
| `4x151936x128` | fp32 | 1 | contiguous | 5607.67 | 5542.50 | 1.012x |
| `2x150x512x512` | fp32 | 1 | channels_last | 5,509.58 | 3,146.67 | 1.751x |
| `2x150x512x512` | bf16 | 1 | channels_last | 3,667.86 | 1,985.69 | 1.847x |
| `8x21x256x256` | fp32 | 1 | channels_last | 1,054.38 | 1,930.98 | 0.546x |
| `8x21x256x256` | bf16 | 1 | channels_last | 591.70 | 996.26 | 0.594x |
| `1x128256` | bf16 | -1 | contiguous | 45.70 | 7.01 | 6.523x |
| `4x151936` | bf16 | -1 | contiguous | 49.82 | 17.57 | 2.835x |
| `64x1000` | bf16 | -1 | contiguous | 19.35 | 8.40 | 2.305x |
| `4096x30522` | bf16 | -1 | contiguous | 2,966.67 | 2,694.35 | 1.101x |

### CODE REVIEW

**Review by malfet — APPROVED** [review]
LGTM, though IMO it would be nice to add 1-2 channels last configs and may be few bf16 ones

**Comment by adityavaid** [comment]
Hi @Isalia20 @malfet  , I came across this PR as I currently am using torch.mps for my work in sglang. We are using custom kernels for sometime for specific op. I would love to add them here as well to keep things simple. Is there a rfc or issue where these kernel additions are being tracked ?

**Comment by Isalia20** [comment]
Appended those numbers in PR description
